tweet_recommendations/other_methods/dbscan_based_method.py

```diff
@@ -0,0 +1,94 @@
+from collections import namedtuple
+from itertools import chain
+from typing import *
+
+import numpy as np
+import pandas as pd
+from scipy.spatial.distance import cosine
+from sklearn.cluster import DBSCAN
+from sklearn.neighbors.kd_tree import KDTree
+
+from ..estimator import Estimator
+
+CentroidData = namedtuple("CentroidData", ["cluster_medoid", "popular_hashtags"])
+
+
+class DBScanBasedEstimator(Estimator):
+    def __init__(self,
+                 verbose: bool = False):
+        self._clusters: np.ndarray = None
+
+        self.clusterizer = None
+        self.neighbours = None
+        self.centroids_data = {}
+
+        self.verbose = verbose
+
+    def fit(self, x: pd.DataFrame, y: Optional[pd.DataFrame] = None, **fit_params):
+        """
+        Builds hashtags representations using w2v method and clusters them using DBScan.
+        :param x: pd.DataFrame with tweet content, user id, and separate hashtags.
+                It is "original_tweets.p" in our case.
+        :param y: None, needed for compatibility.
+        :param fit_params:
+            minimal_hashtag_occurrence: int. If hashtag occurred less than this number
+            then it's not considered during prediction (simply removed).
+            To include all hashtags put number <= 0.
+        :return: self.
+        """
+        minimal_hashtag_occurence = fit_params["minimal_hashtag_occurence"]
+
+        x = self.drop_tweets_with_hashtags_that_occurred_less_than(x, minimal_hashtag_occurence)
+        embeddings = np.array(x["embedding"].to_list())
+
+        print(embeddings.shape)
+
+        self.neighbours = KDTree(embeddings, metric="euclidean")
+        epsilon = np.mean(self.neighbours.query(embeddings, k=2)[0][:, 1])
```

Comment:
Sugerowałbym rozpisać tę linijkę, bo dużo się tutaj dzieje – sam fakt dlaczego `k=2` mógłby być wyjaśniony

Reply:
Z tego, co widzę w paperze, to `k` w tym przypadku powinno być jako parametr konstruktora 
>d for each tweet, we calculate a vector with n elements, where n is the number of a selected set of tweets in the dataset and each vector element is the minimal distance between the tweet and any other tweet. Finally, we use one as the number of minPoints and the average value of this vector as epsilon.

Reply:
`k`, w tym wypadku, to taki lekki hack na poszukiwanie najbliższych sąsiadów. Co do publikacji, to ja to rozumiem tak, że budujemy wektor, nazwijmy go __v__, o liczbie składowych równej _n_:

![](https://latex.codecogs.com/png.latex?%5Cmathbf%7Bv%7D%20%3D%20%5Bd_1%2C%20d_2%2C%20...%2C%20d_n%5D)

gdzie ![](https://latex.codecogs.com/png.latex?d_i) oznacza odległość _i_-tego punktu danych do najbliższego sąsiada. Czyli dla wybranych _n_ punktów szukamy odległości do najbliższego sąsiada, amiright?

Reply:
Inaczej, dla `n` najbliższych sąsiadów wybieramy odległości do nich i uśredniamy te odległości dla argumentu `epsilon`

Reply:
Skoro tak, to trzeba jeszcze przerobić ten kod, bo jest on napisany tak, jak opisałem to w komentarzu wyżej. Nie rozumiem jeszcze skąd bierze się to `n`, bo autorzy nigdzie więcej o nim nie wspominają,

Reply:
Ja nie wiem, czytam ten fragment już po raz 40 i dalej nie wiem o co tu chodzi ...

Reply:
Dobra, empirycznie patrząc na wyniki, n absolutnie musi być hiperparametrem :)